File: packages/stera_recorder/android/src/main/kotlin/open/fpvlabs/stera/ar_recorder/encoding/VideoEncoder.kt
package open.fpvlabs.stera.ar_recorder.encoding

import java.io.File

data class EncoderSetupResult(
    val chosenBitrate: Int,
    val chosenFps: Int,
    val chosenWidth: Int,
    val chosenHeight: Int
)

interface VideoEncoder {
    val actualVideoWidth: Int
    val actualVideoHeight: Int
    val activeVideoBitrate: Int

    fun setup(
        outputFile: File,
        configuredVideoWidth: Int,
        configuredVideoHeight: Int,
        targetEncodeFps: Int,
        targetVideoBitrate: Int,
        fallbackBitrate: Int,
        iFrameIntervalSeconds: Int
    ): EncoderSetupResult?

    fun ensureEncoderSurfaceReady(eglManager: EglManager): Boolean

    fun applyBitrate(newBitrate: Int): Boolean

    fun drain(
        endOfStream: Boolean,
        configuredVideoWidth: Int,
        configuredVideoHeight: Int,
        targetEncodeFps: Int
    )

    fun finalizeStream(
        configuredVideoWidth: Int,
        configuredVideoHeight: Int,
        targetEncodeFps: Int
    )

    fun release()
}

File: packages/stera_recorder/ios/stera_recorder/Sources/stera_recorder/encoding/VideoEncoder.swift
import Foundation
import AVFoundation
import CoreVideo

struct EncoderSetupResult {
    let chosenBitrate: Int
    let chosenFps: Int
    let chosenWidth: Int
    let chosenHeight: Int
}

protocol VideoEncoder {
    var actualVideoWidth: Int { get }
    var actualVideoHeight: Int { get }
    var activeVideoBitrate: Int { get }

    func setup(
        outputPath: String,
        configuredVideoWidth: Int,
        configuredVideoHeight: Int,
        targetEncodeFps: Int,
        targetVideoBitrate: Int
    ) -> EncoderSetupResult?

    func encodeFrame(pixelBuffer: CVPixelBuffer, presentationTime: CMTime) -> Bool
    func finalizeStream()
    func release()
}
